File: tests/harness.py
"""Tiny test harness.

pytest isn't installed in this environment and the report the suite has to
produce - every case with its INPUT and its OUTPUT side by side - isn't
something pytest prints anyway. So the runner is 120 lines of its own.

A test is a function that calls check()/expect() as many times as it likes.
Each call records one row. Rows carry:

    group     which suite it came from
    name      what is being asserted
    inp       the exact input (a URL, a payload, an argument)
    expected  what should have happened
    actual    what did happen
    ok        pass/fail
    severity  how much a failure matters - see SEVERITY

Nothing here is FPL-specific; the suites are.
"""

import json
import time
import traceback

# Ordered worst-first, so a report can sort by it.
SEVERITY = ["critical", "high", "medium", "low", "info"]

ROWS = []
_current = {"group": "?", "severity": "medium"}


def group(name, default_severity="medium"):
    _current["group"] = name
    _current["severity"] = default_severity


def _trim(value, limit=220):
    """Report cells hold one line. Long payloads get elided in the middle so
    both ends stay visible - the end is usually where the interesting part of
    an HTML response or a stack trace is."""
    if not isinstance(value, str):
        try:
            value = json.dumps(value, default=str, ensure_ascii=False)
        except Exception:
            value = repr(value)
    value = " ".join(value.split())
    if len(value) <= limit:
        return value
    keep = (limit - 5) // 2
    return f"{value[:keep]} ... {value[-keep:]}"


def record(name, inp, expected, actual, ok, severity=None, note=""):
    ROWS.append({
        "group": _current["group"],
        "name": name,
        "input": _trim(inp),
        "expected": _trim(expected),
        "actual": _trim(actual),
        "ok": bool(ok),
        "severity": severity or _current["severity"],
        "note": note,
    })
    return ok


def safe(fn, *args, **kwargs):
    """Call fn and return its result, or the exception as a string.

    Needed because the value passed to check() is evaluated before check()
    runs, so a function that raises would escape the harness and abort the
    whole suite instead of recording one failed row."""
    try:
        return fn(*args, **kwargs)
    except Exception as e:
        return f"{type(e).__name__}: {e}"


def check(name, inp, expected_desc, actual_value, predicate, severity=None, note=""):
    """Assert predicate(actual_value). Never raises - a failed case is data,
    not an abort, because the point of the run is the whole table."""
    try:
        ok = bool(predicate(actual_value))
    except Exception as e:
        ok = False
        actual_value = f"{type(e).__name__}: {e}"
    return record(name, inp, expected_desc, actual_value, ok, severity, note)


def expect(name, inp, expected, actual, severity=None, note=""):
    """Equality shorthand."""
    return record(name, inp, expected, actual, actual == expected, severity, note)


def run(suites):
    """Run every suite function, catching anything that escapes so one broken
    suite can't take the report down with it."""
    started = time.time()
    for fn in suites:
        try:
            fn()
        except Exception:
            group(getattr(fn, "__name__", "unknown"))
            record("suite crashed", getattr(fn, "__name__", "?"),
                   "suite runs to completion",
                   _trim(traceback.format_exc(), 400), False, "critical")
    return time.time() - started


def summary():
    total = len(ROWS)
    failed = [r for r in ROWS if not r["ok"]]
    by_group = {}
    for r in ROWS:
        g = by_group.setdefault(r["group"], {"total": 0, "failed": 0})
        g["total"] += 1
        if not r["ok"]:
            g["failed"] += 1
    by_sev = {s: len([r for r in failed if r["severity"] == s]) for s in SEVERITY}
    return {
        "total": total,
        "passed": total - len(failed),
        "failed": len(failed),
        "by_group": by_group,
        "failures_by_severity": by_sev,
    }

File: tests/js/harness.js
/* The JS half of the test harness.
 *
 * Mirrors tests/harness.py deliberately - same row shape, same severities - so
 * the results drop into the one report the Python runner writes rather than
 * being a second set of results somebody has to remember to look at. A test
 * suite nobody runs is worse than no test suite, and a second command to run is
 * how that happens.
 */

'use strict';

const path = require('path');
const fs = require('fs');
const vm = require('vm');

const { createEnvironment } = require('./dom');

const ROWS = [];
let current = { group: '?', severity: 'medium' };

function group(name, severity = 'medium') {
    current = { group: name, severity };
}

function trim(value, limit = 220) {
    let text;
    if (typeof value === 'string') text = value;
    else {
        try { text = JSON.stringify(value); } catch (e) { text = String(value); }
    }
    text = String(text).replace(/\s+/g, ' ');
    if (text.length <= limit) return text;
    const keep = Math.floor((limit - 5) / 2);
    return `${text.slice(0, keep)} ... ${text.slice(-keep)}`;
}

function record(name, input, expected, actual, ok, severity, note) {
    ROWS.push({
        group: current.group,
        name,
        input: trim(input),
        expected: trim(expected),
        actual: trim(actual),
        ok: !!ok,
        severity: severity || current.severity,
        note: note || '',
    });
    return ok;
}

/* Deep equality, so an assertion can be about a whole structure rather than
 * about one field at a time - the same thing `expect` in the Python harness
 * gets for free from ==. */
function same(a, b) {
    if (a === b) return true;
    if (typeof a !== typeof b) return false;
    if (a && b && typeof a === 'object') {
        if (Array.isArray(a) !== Array.isArray(b)) return false;
        const ka = Object.keys(a), kb = Object.keys(b);
        if (ka.length !== kb.length) return false;
        return ka.every(k => same(a[k], b[k]));
    }
    return false;
}

function expect(name, input, expected, actual, severity, note) {
    return record(name, input, expected, actual, same(expected, actual), severity, note);
}

function check(name, input, expectedDesc, actual, predicate, severity, note) {
    let ok;
    try {
        ok = !!predicate(actual);
    } catch (e) {
        ok = false;
        actual = `${e.name}: ${e.message}`;
    }
    return record(name, input, expectedDesc, actual, ok, severity, note);
}

/* Load the site's REAL script files into a fresh stub environment.
 *
 * The shipped files are executed, not copies: a test against a re-typed version
 * of the logic proves only that the copy is self-consistent. `exportNames`
 * becomes an epilogue, because lexical declarations never land on the global
 * object and so cannot be read off the sandbox.
 */
function loadScript(relPaths, { exportNames = [], env = {}, before } = {}) {
    const root = path.resolve(__dirname, '..', '..');
    const paths = Array.isArray(relPaths) ? relPaths : [relPaths];
    // Concatenated into one script so a `const` in an earlier file is in scope
    // for a later one - which is how the browser treats them, and what lets
    // util.js supply esc() to both app.js and compare.js.
    const source = paths
        .map(rel => fs.readFileSync(path.join(root, rel), 'utf8'))
        .join(String.fromCharCode(10) + ';' + String.fromCharCode(10));

    const sandbox = createEnvironment(env);
    if (before) before(sandbox);
    vm.createContext(sandbox);

    const epilogue = exportNames.length
        ? `;globalThis.__exports = { ${exportNames.map(n => `${n}: typeof ${n} !== 'undefined' ? ${n} : undefined`).join(', ')} };`
        : '';
    vm.runInContext(source + String.fromCharCode(10) + epilogue, sandbox,
                    { filename: paths.join(' + ') });
    return { sandbox, exports: sandbox.__exports || {} };
}

/* Let queued promise callbacks run. The scripts under test are full of
 * `.then()` chains, and an assertion made before those have settled is testing
 * the moment before the work happened. */
function flush(times = 6) {
    let p = Promise.resolve();
    for (let i = 0; i < times; i += 1) p = p.then(() => {});
    return p;
}

function summary() {
    const failed = ROWS.filter(r => !r.ok);
    return { total: ROWS.length, passed: ROWS.length - failed.length, failed: failed.length };
}

module.exports = { group, expect, check, record, loadScript, flush, ROWS, summary, same };
